pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,339 @@
+	return reconcile.Result{}, r.setStatusReady(nsTmplSet)
+}
+
+func (r *ReconcileNSTemplateSet) ensureUserNamespaces(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	username := nsTmplSet.GetName()
+
+	// fetch all namespace with owner=username label
+	labels := map[string]string{"owner": username}
+	opts := client.MatchingLabels(labels)
+	userNamespaceList := &corev1.NamespaceList{}
+	if err := r.client.List(context.TODO(), opts, userNamespaceList); err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err, "failed to list namespace with label owner '%s'", username)
+	}
+	userNamespaces := userNamespaceList.Items
+
+	// find next namespace for provisioning namespace resource
+	tcNamespace, userNamespace, found := nextNamespaceToProvision(nsTmplSet.Spec.Namespaces, userNamespaces)
+	if !found {
+		return true, nil
+	}
+
+	// create namespace resource
+	return false, r.ensureNamespace(logger, nsTmplSet, tcNamespace, userNamespace)
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespace(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.NSTemplateSetNamespace, userNamespace *corev1.Namespace) error {
+	username := nsTmplSet.GetName()
+
+	log.Info("provisioning namespace", "namespace", tcNamespace)
+	if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+		return err
+	}
+
+	params := map[string]string{"USER_NAME": username}
+
+	if userNamespace == nil {
+		return r.ensureNamespaceResource(logger, nsTmplSet, tcNamespace, params)
+	}
+	return r.ensureInnerNamespaceResources(logger, nsTmplSet, tcNamespace, params, userNamespace)
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespaceResource(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.NSTemplateSetNamespace, params map[string]string) error {
+	username := nsTmplSet.GetName()
+
+	tmpl, err := r.getTemplateContent(nsTmplSet.Spec.TierName, tcNamespace.Type)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to to retrieve template for namespace type '%s'", tcNamespace.Type)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	objs, err := tmplProcessor.Process(tmpl, params, template.RetainNamespaces)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to process template for namespace type '%s'", tcNamespace.Type)
+	}
+
+	for _, rawObj := range objs {
+		acc, err := meta.Accessor(rawObj.Object)
+		if err != nil {
+			return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "invalid element in template for namespace type '%s'", tcNamespace.Type)
+		}
+
+		// set labels
+		labels := acc.GetLabels()
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels["owner"] = username
+		labels["type"] = tcNamespace.Type
```

Comment:
I can see `tcNamespace.Type` is used multiple times, worth to define it in variable and reuse everywhere

Reply:
I guess, it's minor and either way should be fine.

Reply:
I would vote for using constant too. Same for `revision` label - it's used multiple times too

Reply:
@MatousJobanek Here the point is different.  In case of const vs hard_coded_string is already discussed other place and I have share my point there.  Feel free to open PR with changes you want.